presto-main/src/main/java/com/facebook/presto/sql/planner/GroupingOperationRewriter.java

```diff
@@ -59,34 +60,37 @@ public static Expression rewriteGroupingOperation(GroupingOperation expression,
         // in the group and none of them will be aggregated over. Hence, re-write the
         // GroupingOperation to a constant literal of 0.
         // See SQL:2011:4.16.2 and SQL:2011:6.9.10.
-        boolean isSimpleGroupByPresent = queryNode.getGroupBy().get().getGroupingElements().stream().anyMatch(SimpleGroupBy.class::isInstance);
+        boolean isSimpleGroupByPresent = analysis.getGroupingSets(queryNode).size() == 1;
         if (isSimpleGroupByPresent) {
             return new LongLiteral("0");
         }
         else {
             checkState(groupIdSymbol.isPresent(), "groupId symbol is missing");
 
             Map<Expression, FieldId> columnReferenceFields = analysis.getColumnReferenceFields();
-            List<Expression> columnReferences = ImmutableList.copyOf(analysis.getColumnReferences());
             List<Expression> groupingOrdinals = expression.getGroupingColumns().stream()
-                    .map(groupingColumn -> extractGroupingColumnOrdinal(groupingColumn, columnReferences, columnReferenceFields))
-                    .map(columnOrdinal -> new LongLiteral(Integer.toString(columnOrdinal)))
+                    .peek(groupingColumn -> checkState(
+                            groupingColumn instanceof Identifier || groupingColumn instanceof DereferenceExpression,
+                            "the grouping column is neither an identifier nor a dereference expression"))
+                    .map(columnReferenceFields::get)
+                    .map(FieldId::getFieldIndex)
+                    .map(columnOrdinal -> Integer.toString(columnOrdinal))
+                    .map(LongLiteral::new)
                     .collect(toImmutableList());
 
-            List<List<Expression>> groupingSetOrdinals;
-            ImmutableList.Builder<List<Expression>> groupingSetOrdinalsBuilder = ImmutableList.builder();
-            for (List<Expression> groupingSet : analysis.getGroupingSets(queryNode)) {
-                groupingSetOrdinalsBuilder.add(
-                        groupingSet.stream()
-                            .map(groupingColumn -> extractGroupingColumnOrdinal(groupingColumn, columnReferences, columnReferenceFields))
-                            .map(columnOrdinal -> new LongLiteral(Integer.toString(columnOrdinal)))
-                            .collect(toImmutableList()));
-            }
-            groupingSetOrdinals = groupingSetOrdinalsBuilder.build();
+            List<List<Expression>> groupingSetOrdinals = analysis.getGroupingSets(queryNode).stream()
+                    .map(groupingSet -> groupingSet.stream()
+                            .filter(columnReferenceFields::containsKey)
+                            .map(columnReferenceFields::get)
+                            .map(FieldId::getFieldIndex)
+                            .map(columnOrdinal -> Integer.toString(columnOrdinal))
+                            .map(LongLiteral::new)
+                            .map(Expression.class::cast)
```

Comment:
Minor: you lose type safety here. You can avoid that by any of these
```
.<Expression>map(LongLiteral::new) // instead of preceding line

.<Expression>map(i -> i)

.map(Function.<Expression>identity())
```

Or.. you could capture `FieldId -> Expression` conversion (appears twice: here & above) in a named method and use that as single `.map` step?